### PR TITLE
First-class modules in quotes

### PR DESCRIPTION
**Pull request by jbachurski**
This PR contains:

* Drive-by changes necessary for module (un)packing to work in quotes (until now, only unpacking _mostly_ worked). These are broken down into individual commits. 
* Logic necessary to insert annotations for unannotated packing expressions and unpacking patterns, as these have to be known for a non-HM type inspection. This is the last commit.

There's a test file, `pack_splice_in_constraint.ml`, showing how to trigger some corner case fatal errors right now. These can only happen at metaprogram compilation. As these are tied to of other pieces of work like supporting stage >1 splices (tickets linked), I'd prefer to solve them in separate PRs.

## Review notes

Nothing particularly non-trivial is done to insert the necessary annotations. Module packing is entirely clean, but unpacking is a bit trickier -- the typing logic there is a bit awkward due to the complexities of patterns (particularly GADTs):

* Initially, the typed-tree is constructed unchecked, after which routines like `map_half_typed_cases` might refine the known type, and only eventually check that the unpacking has a package type that is known to be well-formed (e.g. every type is constrained with a closed type expression). 
* We'd really rather construct the `Typedtree` after those checks happen, as that's really when the type is inspected ('determined'). 
* But this seems really annoying to do, so I just keep the type expression as-is in the `Typedtree` and let it flow to be unified later on. Since it is checked to be closed, nothing too surprising should happen to it (unless there's some crazy logic that mutates closed type expressions).

### CODE REVIEW

File: lambda/translquote.ml [hunk 1]
```diff
@@ -2853,8 +2838,9 @@ let type_for_annotation ~env ~loc typ =
         | Tquote ty -> Ttyp_quote (go ty)
         | Tsplice _ ->
           fatal_errorf
-            "Translquote [at %a]:@ Explicitly quantified type variables@ \
-             cannot be spliced@ within quoted higher-rank function types"
+            "Translquote [at %a]:@ Type variables cannot be spliced@ in type \
+             annotations inserted in quotations@ for higher-rank or package \
+             types."
```
Comment (Dreian) [inline lambda/translquote.ml]: Possibly nitpicking, but I don't quite like inserting `@ ` formatting hints (intended to be represented as newlines in most error messages) in places other than near the end of the line.

File: stdlib/camlinternalQuote.ml [hunk 2]
```diff
@@ -1519,8 +1519,9 @@ module Ast = struct
     match pat with
     | PatConstant c when is_negative_const c ->
       pp fmt "(@[%a@])" (print_pat env) pat
-    | PatAny | PatVar _ | PatConstant _ | PatTuple _ | PatUnboxedUnit
-    | PatUnboxedBool _ | PatUnboxedTuple _ | PatVariant (_, None)
+    | PatAny | PatVar _ | PatConstraint _ | PatConstant _ | PatTuple _
+    | PatUnboxedUnit | PatUnboxedBool _ | PatUnboxedTuple _
+    | PatConstruct (_, None) | PatVariant (_, None)
```
Comment (Dreian) [inline stdlib/camlinternalQuote.ml]: The only meaningful change in this diff is the addition of the `PatConstruct (_, None)`  case. Could you try minimising this diff? In particular, I don't think the first line needs to be changed.
(I see the reasoning behind putting all the unboxed cases in a single line, but I'd rather keep the PR clean)

File: testsuite/tests/quotation/build_quotation.ml [hunk 3]
```diff
@@ -807,7 +807,10 @@ module Mod : sig type t = int val mk : 'a -> 'a end
 <[fun (module M : Hashtbl.S) x -> M.clear (M.create x)]>;;
 [%%expect {|
 - : <[(module Hashtbl.S) -> int -> unit]> expr =
-<[fun ((module M) : (module Stdlib.Hashtbl.S)) x -> M.clear (M.create x)]>
+<[
+  fun (((module M) : (module Stdlib.Hashtbl.S)) : (module Stdlib.Hashtbl.S))
+    x -> M.clear (M.create x)
```
Comment (Dreian) [inline testsuite/tests/quotation/build_quotation.ml:850]: I don't think the nested module type constraint is necessary. I believe there is an error in the way `Tpat_constraint` for modules is translated (see related comment in `translquote.ml`).

File: lambda/translquote.ml [hunk 4]
```diff
@@ -2906,6 +2906,12 @@ and quote_pat_extra ~env ~scopes loc pat_lam extra =
       |> quote_core_type ~scopes)
       (Modes.wrap Modes.legacy)
     |> Pat.wrap
+  | Tpat_inspected_type (Module_pack pty) ->
+    Pat.constraint_ loc pat_lam
+      (type_for_annotation ~env ~loc:(to_location loc) pty
+      |> quote_core_type ~scopes)
+      (Modes.wrap Modes.legacy)
+    |> Pat.wrap
```
Comment (Dreian) [inline lambda/translquote.ml:2928]: Related to my comment in the `build_quotation.ml` test (but also relevant for all the other tests): I think we should have a look at what the underlying pattern is in this case; if it is a `Tpat_constraint` and the module type constraint is the same as the one in `pty`, we shouldn't do anything.

This requires changing the type of `quote_pat_extra` (it should take an additional argument).

File: lambda/translquote.ml [hunk 5]
```diff
@@ -3608,6 +3614,14 @@ and quote_expression_extra ~env ~scopes _stage extra lambda =
...
       |> Type_constraint.wrap)
     |> Exp_desc.wrap
+  | Texp_inspected_type (Module_pack pty) ->
+    Exp_desc.constraint_ loc (mk_exp_noattr loc lambda)
+      (Type_constraint.constraint_ loc
+         (type_for_annotation ~env ~loc:(to_location loc) pty
+         |> quote_core_type ~scopes)
+         (Modes.wrap Modes.legacy)
+      |> Type_constraint.wrap)
+    |> Exp_desc.wrap
```
Comment (Dreian) [inline lambda/translquote.ml:3641]: Same comment applies here as with quoting patterns. I don't think we should insert constraints that already exist.

File: testsuite/tests/quotation/typing/inspection/pack.reference [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+Stdlib.ignore
+  ((fun x -> x)
+     (fun ((module M) : (module Pack_types.S)) -> ((module M) : (module
+        Pack_types.S)))
+  : (module Pack_types.S) -> (module Pack_types.S))
+
+Stdlib.ignore
+  ((fun x -> x)
+     (fun (type a) (((module M) : (module Pack_types.S with type t = a)) :
+        (module Pack_types.S with type t = a)) -> M.x)
```
Comment (Dreian) [inline testsuite/tests/quotation/typing/inspection/pack.reference:14]: This constraint duplication appears throughout the tests (in this file and in `build_quotation.ml`). See my comments in `translquote.ml` for `Tpat_inspected_type` and `Texp_inspected_type` for how I think these issues should be addressed.

File: lambda/translquote.ml [hunk 7]
```diff
@@ -2838,8 +2838,9 @@ let type_for_annotation ~env ~loc typ =
         | Tquote ty -> Ttyp_quote (go ty)
         | Tsplice _ ->
           fatal_errorf
-            "Translquote [at %a]:@ Explicitly quantified type variables@ \
-             cannot be spliced@ within quoted higher-rank function types"
+            "Translquote [at %a]:@ Type variables cannot be spliced@ in type \
+             annotations inserted in quotations@ for higher-rank or package \
+             types."
```
Comment (Dreian) [inline lambda/translquote.ml]: I think the `@ ` formatting hints should only appear close to the end of a line.

File: testsuite/tests/quotation/typing/inspection/pack_splice_in_constraint.ml [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                           ^
+Error: The type of this packed module contains variables:
+       "(module S with type t = 'a)"
+|}]
+
+(* Since we erase [$int] into [_] under quotes,
+   the below should fail with an error like the above to be consistent. *)
+(* CR metaprogramming jbachurski: This should fail gracefully.
+   See ticket 7081. *)
+let _ = <[ fun (module M : S with type t = $int) -> M.x ]>
```
Comment (Dreian) [inline testsuite/tests/quotation/typing/inspection/pack_splice_in_constraint.ml]: Could we do something here that isn't `$int`? I believe this is always an invalid type.

**Review by Dreian — CHANGES_REQUESTED** [review]
Well done. The changes are reasonable, but I have a few concerns and suggestions. This mostly involves the "quoting constraints" logic.

The PR description could be more informative, as it's unclear what the status quo is and why this PR is necessary. Could you provide an example of what parts of module packing/unpacking is fixed by this pull request?

File: testsuite/tests/quotation/build_quotation.ml (hunk 3 above)
Comment (jbachurski) [reply testsuite/tests/quotation/build_quotation.ml:850]: Now tracked by ticket 7290, deferred to avoid making the code more complex for now

**Comment by jbachurski** [comment]
I looked into what didn't work before (besides the main subject here -- cross-stage type inspections, which are tested in `pack.ml`, would not result in annotations and you'd see type errors). I think it was just:

* Printing of repeated type constraints with package types. I think I introduced this with a change from another PR,
* The `quote_module_path` function not translating module variables properly, but this has been removed by #6043 in the meantime. 

The specific problem was when the constraints are duplicated (as the generated programs often do, and will sometimes do in the future too): 
```ocaml
<[ fun ((module M : S) : (module S)) -> ((module M : S) : (module S)) ]>;;
[%%expect {|
- : <[(module Hashtbl.S) -> (module Hashtbl.S)]> expr =
<[
  fun ((module M : Stdlib.Hashtbl.S) : (module Stdlib.Hashtbl.S)) ->
    (((module M) : (module Stdlib.Hashtbl.S)) : (module Stdlib.Hashtbl.S))
]>
|}];;
```
It's fine on the pattern (unpacking) but broken on the expression side (packing). In this PR, I opted to always do the verbose thing, but we could later reintroduce the 'flatter' printing of `module M : S` instead of `(module M) : (module S)` once we have better tests.

File: lambda/translquote.ml [hunk 9]
```diff
@@ -2849,8 +2849,8 @@ let type_for_annotation ~env ~loc typ =
         | Tquote ty -> Ttyp_quote (go ty)
         | Tsplice _ ->
           fatal_errorf
-            "Translquote [at %a]: Splices cannot appear in type annotations \
-             inserted in quotations for higher-rank or package types."
+            "Translquote [at %a]:@ Splices cannot appear in type annotations \
+             inserted in quotations@ for higher-rank or package types."
```
Comment (Dreian) [inline lambda/translquote.ml:2857]: Nitpicking: I think `@ ` break hints should roughly match where the newlines in the string are, e.g.:
```
            "Translquote [at %a]: Splices cannot appear in type annotations@ \
             inserted in quotations for higher-rank or package types."
```
Feel free to resolve if you disagree, it's already not followed in this file.

**Review by Dreian — APPROVED** [review]
Looks fine; see if you want to address the one comment.

**Review by Skepfyr — APPROVED** [review]
